conans/client/installer.py

```diff
@@ -36,76 +36,225 @@ def _init_package_info(deps_graph, paths, current_path):
+                    self._out.warn("Filename too long, file excluded: %s" % dest_path)
+            return filtered_files
+
+        if getattr(self._conan_file, 'no_copy_source', False):
+            mkdir(self.build_folder)
+            self._conan_file.source_folder = src_folder
+        else:
+            shutil.copytree(src_folder, self.build_folder, symlinks=True, ignore=check_max_path_len)
+            logger.debug("Copied to %s" % self.build_folder)
+            logger.debug("Files copied %s" % os.listdir(self.build_folder))
+            self._conan_file.source_folder = self.build_folder
+
+        os.chdir(self.build_folder)
+        self._conan_file.build_folder = self.build_folder
+        self._conan_file._conanfile_directory = self.build_folder
+        # Read generators from conanfile and generate the needed files
+        logger.debug("Writing generators")
+        write_generators(self._conan_file, self.build_folder, self._out)
+        logger.debug("Files copied after generators %s" % os.listdir(self.build_folder))
 
-    def report_matches(self):
-        for pattern in self._unused_patterns:
-            self._out.error("No package matching '%s' pattern" % pattern)
+        # Build step might need DLLs, binaries as protoc to generate source files
+        # So execute imports() before build, storing the list of copied_files
+        from conans.client.importer import run_imports
+        copied_files = run_imports(self._conan_file, self.build_folder, self._out)
+
+        try:
+            # This is necessary because it is different for user projects
+            # than for packages
+            logger.debug("Call conanfile.build() with files in build folder: %s"
+                         % os.listdir(self.build_folder))
+            self._out.highlight("Calling build()")
+            with conanfile_exception_formatter(str(self._conan_file), "build"):
+                self._conan_file.build()
+
+            self._out.success("Package '%s' built" % self._conan_file.info.package_id())
+            self._out.info("Build folder %s" % self.build_folder)
+        except Exception as exc:
+            os.chdir(src_folder)
+            self._out.writeln("")
+            self._out.error("Package '%s' build failed" % self._conan_file.info.package_id())
+            self._out.warn("Build folder %s" % self.build_folder)
+            if isinstance(exc, ConanExceptionInUserConanfileMethod):
+                raise exc
+            raise ConanException(exc)
+
+        finally:
+            self._conan_file._conanfile_directory = export_folder
+            # Now remove all files that were imported with imports()
+            for f in copied_files:
+                try:
+                    if f.startswith(self.build_folder):
+                        os.remove(f)
+                except Exception:
+                    self._out.warn("Unable to remove imported file from build: %s" % f)
+
+    @property
+    def build_folder(self):
```

Comment:
Avoids returning the "build_folder" to take into account into the package. Now the class is responsible to know it, and cache it to avoid multiple calls to build_id